File: tokenizer.py
import re
from dataclasses import dataclass
from tokens import BLToken

SPACES_IN_TAB = 2

_token_patterns = [
    (BLToken.NUMBER, r'\d+(\.\d+)?'),
    (BLToken.DEF, r'def\b'),
    (BLToken.IF, r'if\b'),
    (BLToken.ELSE, r'else\b'),
    (BLToken.ELIF, r'elif\b'),
    (BLToken.WHILE, r'while\b'),
    (BLToken.FOR, r'for\b'),
    (BLToken.IN, r'in\b'),
    (BLToken.BREAK, r'break\b'),
    (BLToken.CONTINUE, r'continue\b'),
    (BLToken.RETURN, r'return\b'),
    (BLToken.AND, r'and\b'),
    (BLToken.OR  , r'or\b'),
    (BLToken.NOT, r'not\b'),
    (BLToken.TRUE, r'True\b'),
    (BLToken.FALSE, r'False\b'),
    (BLToken.STRING, r'"[^"]*"'),
    (BLToken.IDENT, r'[A-Za-z_][A-Za-z_0-9]*'),
    (BLToken.PLUS_ASSIGN, r'\+='),
    (BLToken.MINUS_ASSIGN, r'-='),
    (BLToken.STAR_ASSIGN, r'\*='),
    (BLToken.SLASH_ASSIGN, r'/='),
    (BLToken.PERCENT_ASSIGN, r'%='),
    (BLToken.EQUAL, r'=='),
    (BLToken.NEQUAL, r'!='),
    (BLToken.LESS_EQUAL, r'<='),
    (BLToken.GREATER_EQUAL, r'>='),
    (BLToken.PLUS, r'\+'),
    (BLToken.MINUS, r'-'),
    (BLToken.STAR, r'\*'),
    (BLToken.SLASH, r'/'),
    (BLToken.PERCENT, r'%'),
    (BLToken.LESS, r'<'),
    (BLToken.GREATER, r'>'),
    (BLToken.ASSIGN, r'='),
    (BLToken.LPAREN, r'\('),
    (BLToken.RPAREN, r'\)'),
    (BLToken.LBRACK, r'\['),
    (BLToken.RBRACK, r'\]'),
    (BLToken.LBRACE, r'\{'),
    (BLToken.RBRACE, r'\}'),
    (BLToken.COMMA, r','),
    (BLToken.COLON, r':'),
    (BLToken.DOT, r'\.'),
    (BLToken.NEWLINE, r'\n'),
    (BLToken.WHITESPACE, r'[ \t]+'),
    (BLToken.INVALID_TOKEN, r'.+'),
]

_token_re = re.compile('|'.join(f'(?P<{token.name()}>{pattern})' for token, pattern in _token_patterns))
_token_map = {token.name(): token for token, _ in _token_patterns}

@dataclass
class TokenData:
    token: BLToken
    string: str
    lineno: int
    col: int

    def pos(self):
        return [self.lineno, self.col]

class Tokenizer():
    def __init__(self, code: str):
        self.code = code
    
    def count_indent(self, indent: str):
        return len(indent) + indent.count('\t') * (SPACES_IN_TAB - 1)

    def tokenize(self):
        previous_was_newline = False
        comment = False
        indent_stack = [0]
        lineno = 1
        line_start = 0

        for match in re.finditer(_token_re, self.code):
            token_name = match.lastgroup
            token = _token_map[token_name] if token_name else BLToken.INVALID_TOKEN
            value = match.group()

            col = match.start() - line_start + 1

            if token == BLToken.COMMENT:
                comment = True

            if comment and token != BLToken.NEWLINE:
                continue

            comment = False

            if previous_was_newline and token != BLToken.NEWLINE:
                previous_was_newline = False

                if token != BLToken.WHITESPACE:
                    while len(indent_stack) > 1:
                        indent_stack.pop()
                        yield TokenData(BLToken.DEDENT, "", lineno, 1)
                    
                    yield TokenData(token, value, lineno, col)
                    continue

                indent_count = self.count_indent(value)
                if indent_count > indent_stack[-1]:
                    indent_stack.append(indent_count)
                    yield TokenData(BLToken.INDENT, value, lineno, col)
                elif indent_count < indent_stack[-1]:
                    while indent_stack[-1] > indent_count:
                        yield TokenData(BLToken.DEDENT, value, lineno, col)
                        indent_stack.pop()
                    if indent_count != indent_stack[-1]:
                        raise SyntaxError("Inconsistent dedent level")
                continue

            if token == BLToken.NEWLINE:
                previous_was_newline = True
                lineno += 1
                line_start = match.end()

            if token == BLToken.WHITESPACE:
                continue
        
            yield TokenData(token, value, lineno, col)
        
        while len(indent_stack) > 1:
            indent_stack.pop()
            yield TokenData(BLToken.DEDENT, "", lineno, 1)

if __name__ == "__main__":
    with open("samples/example11.bl", "r") as f:
        code = f.read()
    
    print(list(Tokenizer(code).tokenize()))

File: cpp/bootlang/src/tokenizer.cpp
#include <sstream>
#include <regex>
#include <iostream>
#include "tokenizer.hpp"

const size_t SPACES_IN_TABS = 2;

size_t count_indent(std::string &text) {
  size_t count = 0;
  for (char c : text) {
    if (c == ' ') count += 1;
    if (c == '\t') count += SPACES_IN_TABS;
  }

  return count;
}

std::string build_combined_pattern(std::vector<TokenMetadata::RegexPair> regex_pairs) {
  std::string regex;

  for (auto &regex_pair : regex_pairs) {
    if (regex.size() > 0) { regex += "|"; }
    regex += "(" + regex_pair.second + ")";
  }
  
  return regex;
}

Tokenizer::Tokenizer(std::string code) :
  code(std::move(code)) {}

  std::optional<std::vector<TokenData>> Tokenizer::tokenize() {
  std::vector<TokenData> tokens;
  std::vector<size_t> indent_stack{0};

  bool previous_was_newline = false;
  bool comment = false;
  size_t lineno = 1;
  size_t line_start = 0;
  
  static const auto regex_pairs = TokenMetadata::GetInstance().GetTokenRegexes();
  static const auto regex_pattern = build_combined_pattern(regex_pairs);  
  static const std::regex token_regex(build_combined_pattern(regex_pairs), std::regex::ECMAScript);

  auto begin = std::sregex_iterator(code.begin(), code.end(), token_regex);
  auto end = std::sregex_iterator();

  for (auto it = begin; it != end; it++) {
    const std::smatch &match = *it;
    
    Token::Type token = Token::Type::INVALID_TOKEN;
    std::string text;

    for (size_t i = 1; i < match.size(); i++) {
      if (match[i].matched) {
        token = regex_pairs[i - 1].first;
        text = match[i].str();
        break;
      }
    }

    if (token == Token::Type::INVALID_TOKEN) {
      return std::nullopt;
    }

    size_t col = static_cast<size_t>(match.position() - line_start + 1);

    comment = token == Token::Type::COMMENT || comment;
    if (comment && token != Token::Type::NEWLINE) continue;
    comment = false;

    if (previous_was_newline && token != Token::Type::NEWLINE) {
      previous_was_newline = false;

      if (token != Token::Type::WHITESPACE) {
        while (indent_stack.size() > 1) {
          indent_stack.pop_back();
          tokens.push_back(TokenData{Token::Type::DEDENT, "", lineno, 1});
        }

        tokens.push_back(TokenData{token, text, lineno, col});
        continue;
      }

      size_t indent_count = count_indent(text);
      if (indent_count > indent_stack[indent_stack.size() - 1]) {
        indent_stack.push_back(indent_count);
        tokens.push_back(TokenData{Token::Type::INDENT, text, lineno, col});
        continue;
      } 
      
      while (indent_count < indent_stack[indent_stack.size() - 1]) {
        indent_stack.pop_back();
        tokens.push_back(TokenData{Token::Type::DEDENT, text, lineno, col});
      }

      if (indent_count != indent_stack[indent_stack.size() - 1]) {
        return std::nullopt;
      }

      continue;
    }

    if (token == Token::Type::NEWLINE) {
      previous_was_newline = true;
      lineno += 1;
      line_start = match.position() + match.length();
    }

    if (token == Token::Type::WHITESPACE)
      continue;

    tokens.push_back(TokenData{token, text, lineno, col});
  }

  tokens.push_back(TokenData{Token::Type::NEWLINE, "", lineno, 1});
  while (indent_stack.size() > 1) {
    indent_stack.pop_back();
    tokens.push_back(TokenData{Token::Type::DEDENT, "", lineno + 1, 1});
  }

  return tokens;
}